File: Metody Implementacji Algorytmów/L1/hacking-cypher.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int check(string num, int a, int b) {
    int len = num.length();

    int x[len + 1] = {-1};
    int y[len + 1] = {-1};

    x[0] = (num[0] - '0') % a;
    y[len - 1] = (num[len - 1] - '0') % b;

    for (int i = 1; i < len; i++) {
        x[i] = (10 * x[i - 1] + (num[i] - '0')) % a;
    }

    int t = 1;
    for (int i = len - 1; i >= 0; i--) {
        y[i] = (t * (num[i] - '0') + y[i + 1]) % b;
        t = (t * 10) % b;
    }
    for (int i = 0; i < len - 1; i++) {
        if (x[i] == 0 && y[i + 1] == 0 && num[i + 1] != '0') {
            return i;
        }
    }
    return -1;
}

int main() {
    string num;
    cin >> num;
    int a, b;
    cin >> a >> b;

    int i = check(num, a, b);
    if (i != -1) {
        cout << "YES" << endl;
        cout << num.substr(0, i + 1) << endl;
        cout << num.substr(i + 1) << endl;
    } 
    else {
        cout << "NO" << endl;
    }

    return 0;
}

File: Metody Implementacji Algorytmów/L1/hacking-cypher.py
def check(n, a, b):
    x = [-1 for i in range(len(n))]
    y = [-1 for i in range(len(n))]

    x[0] = int(n[0]) % a
    y[len(n) - 1] = int(n[(len(n) - 1)]) % b

    for i in range(1, len(n)):
        x[i] = (10 * x[i - 1] + int(n[i])) % a
    
    t = 10
    for i in range(len(n) - 2, 0, -1):  
        y[i] = (t * int(n[i]) + y[i + 1]) % b
        t *= (10 % b) 

    for i in range(len(n) - 1): 
        if x[i] == 0 and y[i + 1] == 0 and n[i + 1] != '0':
            return i
    return -1

def main():
    num = str(input())
    a, b = map(int, input().split())

    i = check(num, a, b)
    if i != -1:
        print("YES")
        print(num[:i + 1])
        print(num[i + 1:])
    else:
        print("NO")

if __name__ == "__main__":
    main()
